### DISCUSSION TITLE
Adding a Custom Context Propagator

### DISCUSSION THREAD

**Discussion by paychex-ssmithrand** [open-telemetry/opentelemetry-ruby#1355] [Q&A] [answered]
Hi,

I'd like to write a custom propagator to handle the conversion of an in-house proprietary tracing format to OpenTelemetry format, as we start to migrate Ruby-based services. While I see that `OTEL_PROPAGATORS` is an environment variable that can be set, in code there is a discrete list that its looking for: https://github.com/open-telemetry/opentelemetry-ruby/blob/main/sdk/lib/opentelemetry/sdk/configurator.rb#L199

Is there a way to extend the SDK configuration class to accept a custom gem name to use as an additional context propagator? I see that `fetch_propagator` already exists, and could be used in `configure_propagation` to retrieve custom gems if not existent in the hard-coded list.

Thoughts here?

**Comment by ahayworth** [comment, marked as answer]
We don't have a way to automatically "register" a new third party propagator and then make it available by way of environment variable, unfortunately. I think if you came up with a clean way to implement that, we may be open to a PR that implemented the functionality - although I think the SIG might want to discuss a little about that first (here, or in the meeting, etc).

Right now the way you would do this is by setting up the SDK programatically. That would look something like:

```ruby
      my_propagator = My::Custom::Thing.new
      OpenTelemetry::SDK.configure do |c|
        # set up your SDK however you like
        # c.use_all(instrumentation_config)
        # c.resource = resource
        # c.service_name = service_name
        # etc, etc

        # This is the important bit!
        c.propagators = [my_propagator]
      end
```

---

If you choose to implement a registration function, you might have a look at the new `autoprop` package over in opentelemetry-go-contrib. I don't know where the actual docs for it live (it's _very_ new), but:
- implementing PR: https://github.com/open-telemetry/opentelemetry-go-contrib/pull/2258
- function that does the registration: https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/propagators/autoprop/registry.go#L101
- in-code example of its use: https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/propagators/autoprop/example_test.go#L75

That package might be overkill for what we need; perhaps we could do something simpler, somehow.

**Reply by paychex-ssmithrand** [reply]
Thank you for the reply! I couldn't quite glean from the Configurator Class that you could simply use the `propagators` property. That should fit the need for now, and I will think about taking on that enhancement if I am able.

**Reply by ahayworth** [reply]
Fantastic, I'm glad it's helpful! 😄
